pkg/kubelet/kubelet_pods.go

```diff
@@ -1104,82 +1104,90 @@ func (kl *Kubelet) HandlePodCleanups() error {
 	klog.V(3).InfoS("Clean up probes for terminating and terminated pods")
 	kl.probeManager.CleanupPods(runningPods)
 
-	// Terminate any pods that are observed in the runtime but not
-	// present in the list of known running pods from config.
-	runningRuntimePods, err := kl.runtimeCache.GetPods()
-	if err != nil {
-		klog.ErrorS(err, "Error listing containers")
-		return err
-	}
-	for _, runningPod := range runningRuntimePods {
-		switch workerState, ok := workingPods[runningPod.ID]; {
-		case ok && workerState == SyncPod, ok && workerState == TerminatingPod:
-			// if the pod worker is already in charge of this pod, we don't need to do anything
-			continue
-		default:
-			// If the pod isn't in the set that should be running and isn't already terminating, terminate
-			// now. This termination is aggressive because all known pods should already be in a known state
-			// (i.e. a removed static pod should already be terminating), so these are pods that were
-			// orphaned due to kubelet restart or bugs. Since housekeeping blocks other config changes, we
-			// know that another pod wasn't started in the background so we are safe to terminate the
-			// unknown pods.
-			if _, ok := allPodsByUID[runningPod.ID]; !ok {
-				klog.V(3).InfoS("Clean up orphaned pod containers", "podUID", runningPod.ID)
-				one := int64(1)
-				kl.podWorkers.UpdatePod(UpdatePodOptions{
-					UpdateType: kubetypes.SyncPodKill,
-					RunningPod: runningPod,
-					KillPodOptions: &KillPodOptions{
-						PodTerminationGracePeriodSecondsOverride: &one,
-					},
-				})
+	// Only terminate running containers or cleanup runtime resources once we
+	// know our list of known expected pods is complete.
+	//
+	// TODO: Each source should be independently verifiable, since static pods
+	// must be runnable when the apiserver is down and we may need to do cleanup
+	// on their resources even in that scenario.
+	if allSourcesSynced {
```

Comment:
The pod worker is responsible for running a single pod to execution.  When a static pod with the same UID changes on disk, it is not the same pod execution (it's a net new invocation).  The decision was to do it in pod housekeeping because only the housekeeper provides the reconciliation behavior to the pod worker.  The real fix is to make the config loop be a proper reconciler, which this takes a step toward.  I do not think adding it to pod worker will make it more maintainable or more correct because then the pod worker would have to track multiple versions of a pod at once, which is duplicating the responsibility of the pod config loop.